### DISCUSSION TITLE
dag on_start callback

### DISCUSSION THREAD

**Discussion by yoniwg** [apache/airflow#20893] [Q&A]
Does a dag have a callback that is called on the beginning of the dag execution?
I don't see it in the documentation so I guess that it is not a simple one, but what is the best way to achieve it?

My use case is to create a folder for each dag run before the first task execution, so it will be a unique folder per dag-run-id.

Thanks in advance for any idea

**Comment by potiuk** [comment]
Make a task at the beginning of a DAG to create it. Anyway it will only work when you have local executor or shared filesystem as airflow is a distributed filesystem.

**Reply by yoniwg** [reply]
A task is an option, but I want it to be transparent to the user (the DAG writer) so it won't appear as part of the tasks in the UI.
In addition, it require me to parse the task-graph in order to put the task in the right position or to make a sub-DAG from the user DAG... This is quite an overhead for me...
Regarding the shared folder it can work also on a distributed environment if you make an abstract layer of reading/writing to files.
Anyway, it seems that there is no option to execute some operation before the DAG execution, am I correct?

**Reply by potiuk** [reply]
You might want to take a look at https://airflow.apache.org/docs/apache-airflow/stable/concepts/cluster-policies.html. You can write a custom policy that will get fired for all tasks and just before the first task from DAG with a given dag run gets executed, you can do some preparation .

**Reply by potiuk** [reply]
Specifically "tak_instance_mutation" might be particularly useful for your case.

**Comment by Parvathy-Geetha** [comment]
@potiuk Can we use event listeners for this? 
I also have some overheads i want to configure before running certain DAGs and was wondering if that's the way to go. Though event listeners are used across all DAGs, I can probably manipulate the hook definitions to act only on certain DAGs.

**Reply by potiuk** [reply]
Yes. Event listeners are a generic (new) mechanism you can use for that. It has been added recently and while less convenient than writing DAGS (you can only configure it for the whole cluster, and you need to restart all components to change it) if you have cross-DAG initialization/notification etc. listeners are an option
